src/py/flwr/server/state/sqlite_state.py

```diff
@@ -0,0 +1,582 @@
+                # style queries
+                result = rows.fetchall()
+        except KeyError as exc:
+            log(ERROR, {"query": query, "data": data, "exception": exc})
+
+        return result
+
+    def store_task_ins(self, task_ins: TaskIns) -> Optional[UUID]:
+        """Store one TaskIns.
+
+        Usually, the Driver API calls this to schedule instructions.
+
+        Stores the value of the task_ins in the state and, if successful, returns the
+        task_id (UUID) of the task_ins. If, for any reason, storing the task_ins fails,
+        `None` is returned.
+
+        Constraints
+        -----------
+        If `task_ins.task.consumer.anonymous` is `True`, then
+        `task_ins.task.consumer.node_id` MUST NOT be set (equal 0). Any implemenation
+        may just override it with zero instead of validating.
+
+        If `task_ins.task.consumer.anonymous` is `False`, then
+        `task_ins.task.consumer.node_id` MUST be set (not 0)
+        """
+        # Validate task
+        errors = validate_task_ins_or_res(task_ins)
+        if any(errors):
+            log(ERROR, errors)
+            return None
+
+        # Create and set task_id
+        task_id = uuid4()
+        task_ins.task_id = str(task_id)
+
+        # Set created_at
+        created_at: datetime = _now()
+        task_ins.task.created_at = created_at.isoformat()
+
+        # Set ttl
+        ttl: datetime = created_at + timedelta(hours=24)
+        task_ins.task.ttl = ttl.isoformat()
+
+        # Store TaskIns
+        data = (task_ins_to_dict(task_ins),)
+        columns = ", ".join([f":{key}" for key in data[0]])
+        query = f"INSERT INTO task_ins VALUES({columns});"
+
+        self.query(query, data)
+
+        return task_id
+
+    def get_task_ins(
+        self, node_id: Optional[int], limit: Optional[int]
+    ) -> List[TaskIns]:
+        """Get TaskIns optionally filtered by node_id.
+
+        Usually, the Fleet API calls this for Nodes planning to work on one or more
+        TaskIns.
+
+        Constraints
+        -----------
+        If `node_id` is not `None`, retrieve all TaskIns where
+
+            1. the `task_ins.task.consumer.node_id` equals `node_id` AND
+            2. the `task_ins.task.consumer.anonymous` equals `False` AND
+            3. the `task_ins.task.delivered_at` equals `""`.
+
+        If `node_id` is `None`, retrieve all TaskIns where the
+        `task_ins.task.consumer.node_id` equals `0` and
+        `task_ins.task.consumer.anonymous` is set to `True`.
+
+        If `delivered_at` MUST BE set (not `""`) otherwise the TaskIns MUST not be in
+        the result.
+
+        If `limit` is not `None`, return, at most, `limit` number of `task_ins`. If
+        `limit` is set, it has to be greater zero.
+        """
+        if limit is not None and limit < 1:
+            raise AssertionError("`limit` must be >= 1")
+
+        # Retrieve all anonymous Tasks
+        if node_id == 0:
+            return []
```

Comment:
Why do we immediately return here?

`node_id == 0` should never happen, right? It can only be `0` for anonymous nodes, in which case we'd expect the `node_id` to be `None`. I'd suggest raising an `Exception` here, `node_id==0` seems like an error. An empty list suggests that everything went well.

Reply:
True, adjusting it.

Reply:
Let me know what you think about the comment.